File: Trie.java
public class Trie {

    private TrieNode root;

    public Trie() {
        this.root = new TrieNode();
    }

    public void insert(String word) {
        TrieNode current = root;
        for (char character : word.toCharArray()) {
            TrieNode child = current.children.get(character);
            if (child == null) {
                child = new TrieNode();
                current.children.put(character, child);
            }
            current = child;
        }
        current.isWord = true;
    }

    public boolean search(String word) {
        TrieNode current = root;
        for (char character : word.toCharArray()) {
            TrieNode child = current.children.get(character);
            if (child == null) {
                return false;
            }
            current = child;
        }
        return current.isWord;
    }

    public boolean startsWith(String prefix) {
        TrieNode current = root;
        for (char character : prefix.toCharArray()) {
            TrieNode child = current.children.get(character);
            if (child == null) {
                return false;
            }
            current = child;
        }
        return true;
    }

    private class TrieNode {

        private char character;
        private HashMap<Character, TrieNode> children;
        private boolean isWord;

        public TrieNode() {
            this.children = new HashMap<>();
            this.isWord = false;
        }
    }
}
Use code with caution. Learn more
To use the Trie data structure, we can simply create a new instance of the Trie class and call the insert(), search(), and startsWith() methods as needed. For example, the following code inserts the words "hello", "world", and "helloworld" into a trie and then searches for the word "helloworld":

Java
Trie trie = new Trie();

trie.insert("hello");
trie.insert("world");
trie.insert("helloworld");

boolean found = trie.search("helloworld");

System.out.println(found); // true

File: Trie.cpp
#include <unordered_map>
#include <iostream>

struct TrieNode {
    std::unordered_map<char, TrieNode*> children;
    bool is_end_of_word;
    
    TrieNode() : is_end_of_word(false) {}
};

class Trie {
public:
    TrieNode* root;
    
    Trie() { root = new TrieNode(); }
    
    void insert(std::string word) {
        TrieNode* node = root;
        for(char ch : word) {
            if(!node->children.count(ch))
                node->children[ch] = new TrieNode();
            node = node->children[ch];
        }
        node->is_end_of_word = true;
    }
    
    bool search(std::string word) {
        TrieNode* node = root;
        for(char ch : word) {
            if(!node->children.count(ch))
                return false;
            node = node->children[ch];
        }
        return node && node->is_end_of_word;
    }
};

int main() {
    Trie trie;
    trie.insert("hello");
    std::cout << trie.search("hello") << std::endl;  // Output: 1
    std::cout << trie.search("hell") << std::endl;   // Output: 0
    return 0;
}
